usr/src/uts/armv8/boot/boot_uart.c

```diff
@@ -0,0 +1,190 @@
+	 */
+	__asm__ volatile("isb sy":::"memory");
+}
+
+/*
+ * SBSA UART implementation.
+ *
+ * Arm's SystemReady insists that an SBSA UART is implemented, and most
+ * UART implementations are vaguely compatible in that they have compatible
+ * data and flags registers for the bits positions we care about.
+ */
+
+int
+boot_uart_ischar(void)
+{
+	if (boot_uart_mmio_base == NULL)
+		return (0);
+
+	return (!(boot_uart_readreg32(SBSA_UARTFR) & SBSA_UARTFR_RXFE));
+}
+
+int
+boot_uart_getchar(void)
+{
+	if (boot_uart_mmio_base == NULL)
+		return (0);
+
+	while (!boot_uart_ischar())
+		boot_uart_yield();
+
+	return (boot_uart_readreg32(SBSA_UARTDR) & 0xff);
+}
+
+void
+boot_uart_putchar(int c)
+{
+	if (boot_uart_mmio_base == NULL)
+		return;
+
+	while (boot_uart_readreg32(SBSA_UARTFR) & SBSA_UARTFR_TXFF)
+		boot_uart_yield();
+
+	if ((c & 0xff) == '\n')
+		boot_uart_putchar('\r');
+
+	boot_uart_writereg32(SBSA_UARTDR, c & 0xff);
+	while (!(boot_uart_readreg32(SBSA_UARTFR) & SBSA_UARTFR_TXFE))
+		boot_uart_yield();
+}
+
+/*
+ * Boot services initialisation
+ */
+void
+boot_uart_init(struct xboot_info *xbp)
+{
+#if (defined(EARLY_UART_PA) && EARLY_UART_PA > 0) && \
+    (defined(EARLY_UART_TYPE) && EARLY_UART_TYPE != XBI_BSVC_UART_NONE)
+	if (xbp == NULL ||
+	    xbp->bi_bsvc_uart_mmio_base == 0 ||
+	    xbp->bi_bsvc_uart_type == XBI_BSVC_UART_NONE) {
+		boot_uart_mmio_base = (caddr_t)(EARLY_UART_PA);
+		boot_uart_type = (xbi_bsvc_uart_type_t)(EARLY_UART_TYPE);
+
+		switch (boot_uart_type) {
+		case XBI_BSVC_UART_PL011:	/* fallthrough */
+		case XBI_BSVC_UART_SBSA2X:	/* fallthrough */
+		case XBI_BSVC_UART_SBSA:	/* fallthrough */
+		case XBI_BSVC_UART_BCM2835:
+			break;
+		default:
+			boot_uart_mmio_base = NULL;
+			boot_uart_type = XBI_BSVC_UART_NONE;
+			break;
+		}
+
+		return;
+	}
+#endif
+
+	if (xbp == NULL)
+		return;
+
+	if (xbp->bi_bsvc_uart_mmio_base == 0) {
+#if (defined(EARLY_UART_PA) && (EARLY_UART_PA > 0))
+		boot_uart_mmio_base = (caddr_t)(EARLY_UART_PA);
+#endif
+	} else {
+		boot_uart_mmio_base = (caddr_t)(xbp->bi_bsvc_uart_mmio_base);
+	}
+
+	if (boot_uart_mmio_base == NULL)
+		return;
+
+	if (xbp->bi_bsvc_uart_type == XBI_BSVC_UART_NONE) {
+#if (defined(EARLY_UART_TYPE) && (EARLY_UART_TYPE != XBI_BSVC_UART_NONE))
+		boot_uart_type = (xbi_bsvc_uart_type_t)(EARLY_UART_TYPE);
+#endif
+	} else {
+		boot_uart_type = xbp->bi_bsvc_uart_type;
+	}
+
+	switch (boot_uart_type) {
+	case XBI_BSVC_UART_PL011:	/* fallthrough */
+	case XBI_BSVC_UART_SBSA2X:	/* fallthrough */
+	case XBI_BSVC_UART_SBSA:	/* fallthrough */
+	case XBI_BSVC_UART_BCM2835:
+		break;
+	default:
+		boot_uart_mmio_base = NULL;
+		boot_uart_type = XBI_BSVC_UART_NONE;
+		break;
+	}
+}
+
+#if !defined(_BOOT)
+/*
+ * Boot services relocation
+ *
+ * Called after kernel memory is up and running to relocate the boot services
+ * UART base address to the device arena.
+ *
+ * XXXARM: we should not need this... there's an (unused) switcheroo in i86pc
```

Comment:
The x86 switcheroos and kbm are a thing I haven't figured out yet, either

Reply:
For the prom I/O redirection, the switch is in `kadb_uses_kernel`, which is documented as totally misnamed. This is called from dynamic_console_config via consconfig_init_input.

`kbm` is going to take a fair bit of untangling.